pennylane/plugins/expt_tensornet.py

```diff
@@ -0,0 +1,197 @@
+#  device
+#========================================================
+
+
+class TensorNetwork(Device):
+    """Experimental Tensor Network simulator device for PennyLane.
+
+    Args:
+        wires (int): the number of modes to initialize the device in
+    """
+    name = 'PennyLane TensorNetwork simulator plugin'
+    short_name = 'expt.tensornet'
+    pennylane_requires = '0.7'
+    version = '0.7.0'
+    author = 'Xanadu Inc.'
+    _capabilities = {"model": "qubit", "tensor_observables": True}
+
+    _operation_map = {
+        'QubitStateVector': None,
+        'BasisState': None,
+        'QubitUnitary': unitary,
+        'PauliX': X,
+        'PauliY': Y,
+        'PauliZ': Z,
+        'Hadamard': H,
+        'S': S,
+        'T': T,
+        'CNOT': CNOT,
+        'SWAP': SWAP,
+        'CSWAP':CSWAP,
+        'CZ': CZ,
+        'PhaseShift': Rphi,
+        'RX': Rotx,
+        'RY': Roty,
+        'RZ': Rotz,
+        'Rot': Rot3,
+        'CRX': CRotx,
+        'CRY': CRoty,
+        'CRZ': CRotz,
+        'CRot': CRot3
+    }
+
+    _observable_map = {
+        'PauliX': X,
+        'PauliY': Y,
+        'PauliZ': Z,
+        'Hadamard': H,
+        'Hermitian': hermitian,
+        'Identity': identity
+    }
+
+    def __init__(self, wires):
+        super().__init__(wires, shots=1)
+        self.eng = None
+        self.analytic = True
+        self._nodes = []
+        self._edges = []
+        self._zero_state = np.zeros([2] * wires)
+        self._zero_state[tuple([0] * wires)] = 1.0
+        # TODO: since this state is separable, can be more intelligent about not making a dense matrix
+        self._state = self._add_node(self._zero_state, wires=tuple(w for w in range(wires)), name="AllZeroState")
+        self._free_edges = [e for e in self._state.edges]  # we need this list to be distinct from self._state.edges
+
+    def _add_node(self, A, wires, name="UnnamedNode"):
+        name = "{}{}".format(name, tuple(w for w in wires))
+        if isinstance(A, tn.Node):
+            A.set_name(name)
+            node = A
+        else:
+            node = tn.Node(A, name=name)
+        self._nodes.append(node)
+        return node
+
+    def _add_edge(self, node1, idx1, node2, idx2):
+        edge = tn.connect(node1[idx1], node2[idx2])
+        self._edges.append(edge)
+        return edge
+
+    def pre_apply(self):
+        self.reset()
+
+    def apply(self, operation, wires, par):
+        A = self._get_operator_matrix(operation, par)
+        num_mult_idxs = len(wires)
+        A = np.reshape(A, [2] * num_mult_idxs * 2)
+        op_node = self._add_node(A, wires=wires, name=operation)
+        for idx, w in enumerate(wires):
+            self._add_edge(op_node, num_mult_idxs + idx, self._state, w)
+            self._free_edges[w] = op_node[idx]
+        # TODO: can be smarter here about collecting contractions?
+        self._state = tn.contract_between(op_node, self._state, output_edge_order=self._free_edges)
```

Comment:
Ah, so with every apply, we contract so that the state becomes a single node?

Reply:
Yes, that's how it's currently implemented, but there are smarter ways to do this:

- delegate all contractions until the end (but prevents there being a `_state` attribute. Instead, a user calling `_state` would need to hit a function which actually computes the state
- since the initial state is unentangled, start off with separate nodes for each subsystem